### PR TITLE
refactor and cleanup region errors for NLL

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a WIP commit. It simplifies some of the code from https://github.com/rust-lang/rust/pull/51536 and extends a few more steps towards the errors that @davidtwco and I were shooting for. These are intended as a replacement for the general "unable to infer lifetime" messages -- one that is actually actionable. We're certainly not there yet, but the overall shape hopefully gets a bit clearer.

I'm thinking about trying to open up an internals thread to sketch out the overall plan and perhaps discuss how to get the wording right, which special cases to handle, etc.

r? @estebank 
cc @davidtwco

### CODE REVIEW

**Comment by nikomatsakis** [comment]
That said, I might like to land this PR anyway, just to give us more to build from.

**Comment by nikomatsakis** [comment]
I haven't updated the stderr files yet, but the general output we are creating looks like this:

```
error: unsatisfied lifetime constraints
  --> borrowck/issue-45983.rs:17:18
   |
16 |     let x = None;
   |         - region `'2` appears in the type of `x`
17 |     give_any(|y| x = Some(y));
   |               -  ^^^^^^^^^^^ free region requires that `'1` must outlive `'2`
   |               |
   |               region `'1` appears in this argument
```

There are a bunch of "to do" items here:

- In this particular case, I think we should add some special wording about closures:
  - something to indicate that the reference `y` cannot escape the closure body
- The "region `'2` appears in  the type of `x`" messages are *meant* to display a "fully elaborated" type, like "the fully elaborated type of `x` is `Option<&'2 usize>`", which should help users to pinpoint otherwise anonymous regions (same for the one about `'1`)

**Comment by nikomatsakis** [comment]
Another example:

```
error: unsatisfied lifetime constraints
  --> underscore-lifetime/dyn-trait-underscore.rs:18:5
   |
16 | fn a<T>(items: &[T]) -> Box<dyn Iterator<Item=&T>> {
   |         ----- region `'1` appears in this argument
17 |     Box::new(items.iter()) //~ ERROR cannot infer an appropriate lifetime
   |     ^^^^^^^^^^^^^^^^^^^^^^ cast requires that `'1` must outlive `'static`
```

In this case, my final goal might look like this:

```
error: unsatisfied lifetime constraints
  --> underscore-lifetime/dyn-trait-underscore.rs:18:5
   |
16 | fn a<T>(items: &[T]) -> Box<dyn Iterator<Item=&T>> {
   |         -----                          ----------------------------- lifetime bound defaults to `'static`
   |         |
   |         the fully elaborated type of `items` is `&'1 [T]`
17 |     Box::new(items.iter()) //~ ERROR cannot infer an appropriate lifetime
   |     ^^^^^^^^^^^^^^^^^^^^^^ cast requires that `'1` must outlive `'static`
```

This could be done with:

- the same 'fully elaborated types' fixme
- a special case that looks for the case where the `'static` arises from an object type in the return type
    - I *think* we have all the info we need to see that...

**Comment by estebank** [comment]
Let me know if you wish to merge _as is_ or with a mild clean up and I can give it a further pass once they're in master.

**Comment by nikomatsakis** [comment]
@estebank 

> Let me know if you wish to merge as is or with a mild clean up and I can give it a further pass once they're in master.

Mostly I don't want to try and get the "perfect message" in this PR. I'd be happy to do some mild cleanup, but also happy to land as is and do cleanup after landing.

Also, this code will hopefully pass travis now. (Had a few things to fix.)

**Comment by nikomatsakis** [comment]
@estebank I opened [this thread on internals for discussion](https://internals.rust-lang.org/t/presenting-lifetime-inference-errors-better-in-nll/7881)

File: src/test/ui/impl-trait/static-return-lifetime-infered.nll.stderr [hunk 1]
```diff
@@ -10,17 +10,19 @@ warning: not reporting region error due to nll
...
 
-error: free region `` does not outlive free region `'static`
+error: unsatisfied lifetime constraints
   --> $DIR/static-return-lifetime-infered.rs:17:9
    |
+LL |     fn iter_values_anon(&self) -> impl Iterator<Item=u32> {
+   |                         ----- lifetime `'1` appears in this argument
 LL |         self.x.iter().map(|a| a.0)
-   |         ^^^^^^^^^^^^^
+   |         ^^^^^^ cast requires that `'1` must outlive `'static`
```
Comment (estebank) [inline src/test/ui/impl-trait/static-return-lifetime-infered.nll.stderr]: I'm intrigued where the `cast` category is coming from here...

**Review by estebank — APPROVED** [review]

**Pull request merged**